lib/modules.nix

```diff
@@ -87,31 +90,70 @@ rec {
       result = { inherit options config; };
     in result;
 
+  # collectModules :: (modulesPath: String) -> (modules: [ Module ]) -> (args: Attrs) -> [ Module ]
+  #
+  # Collects all modules recursively through `import` statements, filtering out
+  # all modules in disabledModules.
+  collectModules = let
 
- # Filter disabled modules. Modules can be disabled allowing
- # their implementation to be replaced.
- filterModules = modulesPath: modules:
-   let
-     moduleKey = m: if isString m then toString modulesPath + "/" + m else toString m;
-     disabledKeys = map moduleKey (concatMap (m: m.disabledModules) modules);
-   in
-     filter (m: !(elem m.key disabledKeys)) modules;
-
-  /* Close a set of modules under the ‘imports’ relation. */
-  closeModules = modules: args:
-    let
-      toClosureList = file: parentKey: imap1 (n: x:
-        if isAttrs x || isFunction x then
-          let key = "${parentKey}:anon-${toString n}"; in
-          unifyModuleSyntax file key (applyIfFunction key x args)
-        else
-          let file = toString x; key = toString x; in
-          unifyModuleSyntax file key (applyIfFunction key (import x) args));
-    in
-      builtins.genericClosure {
```

Comment:
So it looks like `filterAttrs` wasn't the problem.
Perhaps `genericClosure` speeds things up in the old implementation?

The new implementation still computes a closure; just on a smaller relation. Did you try implementing it with `genericClosure`?

Reply:
`genericClosure` was used it because it deduplicates modules occuring multiple times with the same key automatically which was possible in the old implementation because it doesn't matter how many times a module occured.

However now with a recursive `disabledModules` it does, because depending on how many times it occurs, it should be disabled or not.

So instead of using `genericClosure` to transform `list of modules -> list of all modules recursively`, I use `collectStructuredModules` to extract the full structure of the imports recursively, then filtering out subtrees that are disabled by key, and only then deduplicating multiple same keys.

Reply:
I don't think deduplication was the only reason to use `genericClosure`.
In both the old and the new implementation, we can rely on computing the `disabledKeys` once. 

The old implementation applied `disabledKeys` too late; after forgetting the import graph by computing the closure.

Your new code uses `disabledModules` during the closure computation (`flattenKeys`).
This seems to be equivalent applying `disabledKeys` *before* the closure computation. If I understand correctly, what we want is the `closure of ( imports minus disabled )`, rather than `( closure of imports ) minus disabled`.

So it seems like `genericClosure` can actually be used in your new code, by filtering out disabled modules in the `operator`.

If I got this completely wrong, please clarify how it should be decided that a module is disabled or not depending on how many times it occurs.

Reply:
Ah yes I see what you mean now. The problem with that is that `disabledModules` can appear arbitrarily deep in the import graph. As a controversial example, a module `foo.nix` could contain

```nix
{
  imports = [
    { disabledModules = [ "foo.nix" ]; }
  ];
}

We can't know about the fact that `foo.nix` is disabled until we loaded its imports.

Reply:
Yes, it will still need two passes, one to gather the `disabledModules` and then another to compute the (smaller) closure.

My suggestion only optimizes roughly the `flattenKeys` code.

Reply:
So the first "pass" is unchanged and is not really a closure computation. It's just the `disabledKeys`.

Reply:
Oh I see, hmm that might work, though I think it would require loading all files twice. I might give that a try though.

Reply:
Tried this out in https://github.com/Infinisil/nixpkgs/commit/d686167ea49e9c8a3689c24fecd6e6b3e34951ae, though the performance is worse as I would've expected: Over two runs, the `tested` set took 238.3s on average to evaluate.

Reply:
Um, apparently my machine is just really slow right now for some reason, as it took also about 240s for the version of this PR right now :/

Reply:
Got myself a powerful remote machine just to benchmark this, I'll report again once I have the numbers

Reply:
Stats for the two approaches

| Approach | N | min | max | mean | stddev | sderr |
| --- | --- | --- | --- | --- | --- | --- |
| Current | 9 | 289.838 | 302.766 | 295.92 | 3.80687 | 1.26896 |
| `genericClosure` | 9 | 290.656 | 306.213 | 298.174 | 4.72164 | 1.57388 |

Sooo it doesn't seem to matter much, but the current one is slightly better on average. Also it's much nicer: Less duplicated code, doesn't import files twice, does a single reading of the structure.

Reply:
I don't understand why you have to load the modules twice. `flattenKeys` looks like a `genericClosure` to me. Something like:

```
includedKeysClosure = builtins.genericClosure {
  startSet = lib.mapAttrsToList (key: value: { inherit key value; }) keys;
  operator = {key, value}: builtins.removeAttrs value.keys disabledKeys;
};
includedModules = map (x: x.value.module) includedKeysClosure;
#                           ^ x.value corresponds to flattenKeys keys
```

Reply:
Oh I see, will try that too then

Reply:
Yup that's better:

| Approach | N | min | max | mean | sd | sderr |
| --- | --- | --- | --- | --- | --- | --- |
| Before PR | 9 | 264.469 | 272.741 | 268.26 | 2.83212 | 0.944039 |
| Current PR | 9 | 270.456 | 281.266 | 275.199 | 3.39928 | 1.13309 |
| Alternative | 9 | 264.811 | 278.93 | 272.494 | 3.96077 | 1.32026 |

Reply:
Was able to avoid both a `listToAttrs` and an `mapAttrsToList` by just keeping everything as lists.